### PR TITLE
Add support for asynchronous assertions through logging

### PR DESCRIPTION
**Pull request by Avaq**
Very much a work in progress. I've described the essence of the idea in a comment.

### CODE REVIEW

**Comment by Avaq** [comment]
The implementation changed slightly since last time. Now I collect all "output", including log calls, into a single list during parsing. That whole list is injected as the output thunk during rewriting. The evaluator will be updated to consider not just one, but all output channels during evaluation.

**Comment by Avaq** [comment]
The latest commit is the result of some high focus debugging. All tests but the coffeescript and commonjs are passing now. Coffeescript not passing is expected. I'm not sure why commonjs tests are not passing - they all generate the same error: `require is not defined`.

File: lib/doctest.js [hunk 1]
```diff
@@ -322,20 +346,25 @@ function wrap$js(test, sourceType) {
...
       '__doctest.enqueue({',
       '  type: "' + OUTPUT + '",',
       '  ":": ' + test[OUTPUT].loc.start.line + ',',
       '  "!": ' + test['!'] + ',',
       '  thunk: function() {',
-      '    return ' + test[OUTPUT].value + ';',
+      '    return [\n' + test[OUTPUT].map (function(out) {
+        return '      ' +
+               '{channel: ' + JSON.stringify (out.channel) +
+               ', value: (\n' + out.value + '\n      )}';
```
Comment (Avaq) [inline lib/doctest.js]: This approach with the newlines is to allow users to append comments to their input and output lines. It does, however, mean that they can no longer append semicolons to their expressions. I'm not sure if that was allowed.

Comment (Avaq) [reply lib/doctest.js]: If we want to support expressions with semicolons, we'll have to use an IIFE here instead.

**Comment by Avaq** [comment]
Hm, there might be a little more work to do before commonjs, coffeescript, and amd modules have the logging feature. I think I'll find out when writing the tests.

File: .travis.yml [hunk 2]
```diff
@@ -1,8 +1,6 @@
 sudo: false
 language: node_js
 node_js:
-  - "8"
-  - "9"
```
Comment (davidchambers) [inline .travis.yml:5]: :thumbsup:

File: lib/common.js [hunk 3]
```diff
@@ -23,13 +23,17 @@ exports.runDoctests = function(doctest, program) {
...
     process.exit (1);
   }
-  Promise.all (program.args.map (function(path) {
-    return (doctest (path, program)).then (function(results) {
-      return results.reduce (function(status, tuple) {
-        return tuple[0] ? status : 1;
-      }, 0);
+  program.args.reduce (function(promise, path) {
+    return promise.then (function(statuses) {
+      return doctest (path, program).then (function(results) {
```
Comment (davidchambers) [inline lib/common.js:28]: I slightly prefer the following:

```suggestion
      return (doctest (path, program)).then (function(results) {
```

File: lib/common.js [hunk 4]
```diff
@@ -23,13 +23,17 @@ exports.runDoctests = function(doctest, program) {
...
+        var status = results.reduce (function(status, tuple) {
+          return tuple[0] ? status : 1;
+        }, 0);
```
Comment (davidchambers) [inline lib/common.js:31]: I realize this is existing code, but I think it should be rewritten in terms of `Array#every`:

```suggestion
        var status = results.every (function(t) { return t[0]; }) ? 0 : 1;
```

File: lib/common.js [hunk 5]
```diff
@@ -23,13 +23,17 @@ exports.runDoctests = function(doctest, program) {
...
+        return statuses.concat ([status]);
+      });
     });
-  })).then (function(statuses) {
+  }, Promise.resolve ([]))
+  .then (function(statuses) {
     process.exit (statuses.every (function(s) { return s === 0; }) ? 0 : 1);
```
Comment (davidchambers) [inline lib/common.js:37]: Could we simply keep track of a single status?

File: lib/doctest.js [hunk 6]
```diff
@@ -200,17 +203,39 @@ function processLine(
...
       $test = accum.tests[accum.tests.length - 1];
-      $test[accum.state = OUTPUT] = {value: value};
-      output ($test);
+      $test[accum.state = OUTPUT].values.push ({
+        channel: null,
+        value: value
+      });
+      if ($test[OUTPUT].values.length === 1) {
+        output ($test);
+      }
```
Comment (davidchambers) [inline lib/doctest.js:238]: What is the significance of `$test[OUTPUT].values` containing *exactly* one element? Will the array ever contain more than one element?

File: lib/doctest.js [hunk 7]
```diff
@@ -322,20 +347,28 @@ function wrap$js(test, sourceType) {
          type === 'ImportDeclaration' ||
          type === 'VariableDeclaration' ?
     test[INPUT].value :
+    // TODO: Why are we enqueing two things if it could just be one structure?
```
Comment (davidchambers) [inline lib/doctest.js:350]: I believe we are enqueuing one or two things depending on whether the input has associated output.

File: lib/doctest.js [hunk 8]
```diff
@@ -523,50 +562,113 @@ function commonjsEval(source, path) {
...
-function run(queue) {
-  return queue.reduce (function(accum, io) {
-    var thunk = accum.thunk;
-    if (io.type === INPUT) {
-      if (thunk != null) thunk ();
-      accum.thunk = io.thunk;
-    } else if (io.type === OUTPUT) {
-      var either;
+function run(options, queue, optionalLogMediator) {
+  var logMediator = optionalLogMediator || {};
```
Comment (davidchambers) [inline lib/doctest.js:569]: Could this be moved to the call sites? In other words, could we *require* a log mediator?

File: lib/doctest.js [hunk 9]
```diff
@@ -523,50 +562,113 @@ function commonjsEval(source, path) {
...
-          '! ' + expected.name + expected.message.replace (/^(?!$)/, ': ') :
-          show (expected),
-        io[':']
-      ]);
-    }
-    return accum;
-  }, {results: [], thunk: null}).results;
+      return typeof options.logFunction === 'string' ?
+             (awaitOutput ()).then (verifyOutput) :
+             Promise.resolve (verifyOutput ());
```
Comment (davidchambers) [inline lib/doctest.js:616]: Does the logging functionality need to be optional? Are you worried about naming collisions?

File: lib/doctest.mjs [hunk 10]
```diff
@@ -22,35 +22,41 @@ export default async function(path, options) {
...
 }
 
-function wrap(source) {
+function wrap(source, logFunction) {
   return common.unlines ([
     'export const __doctest = {',
     '  queue: [],',
-    '  enqueue: function(io) { this.queue.push(io); }',
+    '  enqueue: function(io) { this.queue.push(io); },',
+    '  logMediator: {emit: function(){}}',
```
Comment (davidchambers) [inline lib/doctest.mjs:48]:
```suggestion
    '  logMediator: {emit: function() {}}',
```

File: lib/doctest.mjs [hunk 11]
```diff
@@ -22,35 +22,41 @@ export default async function(path, options) {
...
     '};',
-    '',
-    source
-  ]);
+    ''
+  ]) + (logFunction != null ? common.unlines ([
+    'const ' + logFunction + ' = channel => value => {',
+    '  __doctest.logMediator.emit ({channel, value});',
+    '};',
```
Comment (davidchambers) [inline lib/doctest.mjs:54]: I think this is our first use of arrow functions. Should we also use `() => {}` in place of `function() {}` a few lines above?

File: lib/doctest.mjs [hunk 12]
```diff
@@ -22,35 +22,41 @@ export default async function(path, options) {
...
+    ''
+  ]) : []) + (source);
```
Comment (davidchambers) [inline lib/doctest.mjs:56]:
```suggestion
  ]) : []) + source;
```

File: test/index.js [hunk 13]
```diff
@@ -255,15 +255,21 @@ testCommand ('bin/doctest --print test/commonjs/exports/index.js', {
...
     '__doctest.enqueue({',
     '  type: "output",',
     '  ":": 2,',
     '  "!": false,',
     '  thunk: function() {',
-    '    return 42;',
+    '    return [',
+    '      {channel: null, value: (',
+    '        42',
+    '      )}',
```
Comment (davidchambers) [inline test/index.js:271]: This is fine, although I would prefer one of the following styles:

```javascript
{channel: null, value: (42)}
```

```javascript
{
  channel: null,
  value: (
    42
  )
}
```

File: lib/doctest.js (hunk 6 above)
Comment (Avaq) [reply lib/doctest.js:238]: It contains additional elements if the user specifies expected log lines. Eg.

```js
> ( log ('stdout') (1)
. , setTimeout (log ('stdout'), 1, 2)
. , log ('stderr') (3) )
[stdout]: 1
[stderr]: 3
undefined
[stdout]: 2
```

Here there's 4 elements collected into the outputs array.

File: lib/doctest.js (hunk 7 above)
Comment (Avaq) [reply lib/doctest.js:350]: I think it'd be more neat if the input and output would be enqueued as a single structure.

File: lib/doctest.js (hunk 8 above)
Comment (Avaq) [reply lib/doctest.js:569]: This might be a leftover from an earlier version where I didn't yet pass `options` down, and the absence of the logMediator would indicate that logging is not enabled.

File: lib/doctest.js (hunk 9 above)
Comment (Avaq) [reply lib/doctest.js:616]: Yes. The user explicitly injects a log function with a given name, which must not collide with other values defined in their module.

File: lib/doctest.mjs (hunk 11 above)
Comment (Avaq) [reply lib/doctest.mjs:54]: Oh. Perhaps I should use a regular function.

File: test/index.js (hunk 13 above)
Comment (Avaq) [reply test/index.js:271]: The reason for the newlines is to allow a user to inject `42 // Expect forty-two` there. Without the newlines that would cause the closing bracket to be commented out.

File: lib/doctest.js (hunk 6 above)
Comment (davidchambers) [reply lib/doctest.js:238]: What would the value of `$test[OUTPUT].values` be in the example above?

File: lib/doctest.js (hunk 7 above)
Comment (davidchambers) [reply lib/doctest.js:350]: Yes, that would be neater. :)

File: lib/doctest.js (hunk 9 above)
Comment (davidchambers) [reply lib/doctest.js:616]: You want to save users from resorting to `--log-function asdf`, I assume. :joy:

File: test/index.js (hunk 13 above)
Comment (davidchambers) [reply test/index.js:271]: I see. Could we use the second of my preferred styles?

File: lib/doctest.js (hunk 8 above)
Comment (Avaq) [reply lib/doctest.js:574]: Ah, no. It's because some logic hasn't yet been updated to supply a logMediator. This code should go away once all source types implement logging.

File: test/index.js (hunk 13 above)
Comment (Avaq) [reply test/index.js:293]: Ideally we'd use an immediately invoked implicitly returned arrow function, because it would allow for semicolon injection as well as comments.

File: lib/common.js [hunk 14]
```diff
@@ -23,14 +23,15 @@ exports.runDoctests = function(doctest, program) {
...
-  Promise.all (program.args.map (function(path) {
-    return (doctest (path, program)).then (function(results) {
-      return results.reduce (function(status, tuple) {
-        return tuple[0] ? status : 1;
-      }, 0);
+  program.args.reduce (function(promise, path) {
+    return promise.then (function(ok) {
+      return (doctest (path, program)).then (function(results) {
+        return ok && results.every (function(t) { return t[0]; });
+      });
```
Comment (davidchambers) [inline lib/common.js:30]: :thumbsup:

File: lib/doctest.js [hunk 15]
```diff
@@ -197,20 +212,46 @@ function processLine(
...
-    } else if (accum.state === INPUT) {
+      $test[OUTPUT].values[$test[OUTPUT].values.length - 1].value += '\n'
+                                                                  + value;
+      appendToOutput ($test);
+    } else if (MATCH_LOG.test (trimmedLine)) {
+      accum.state = LOG;
+      value = stripLeading (1, ' ', trimmedLine.replace (MATCH_LOG, ''));
+      $test = accum.tests[accum.tests.length - 1];
+      $test[OUTPUT].values.push ({
+        channel: MATCH_LOG.exec (trimmedLine)[1],
```
Comment (davidchambers) [inline lib/doctest.js:238]: I'm not totally happy with this (guarded) unsafe expression. I would prefer this approach:

```javascript
var match;
if (...) {
  // ...
} else if ((match = MATCH_LOG.exec (trimmedLine)) != null) {
  // ...
  // match[1]
  // ...
}
```

File: lib/doctest.js [hunk 16]
```diff
@@ -500,73 +555,169 @@ function rewrite$coffee(options, input) {
...
-function functionEval(source) {
+function functionEval(options, source, logFunction) {
   //  Functions created via the Function function are always run in the
   //  global context, which ensures that doctests can't access variables
   //  in _this_ context.
   //
-  //  The `evaluate` function takes one argument, named `__doctest`.
-  var evaluate = Function ('__doctest', source);
+  //  The `evaluate` function takes one or two arguments, `__doctest` is
+  //  the doctest runtime broker, and optionally it takes a logger function.
```
Comment (davidchambers) [inline lib/doctest.js:564]: Could we require the logging function here too?

Comment (Avaq) [reply lib/doctest.js:564]: The user decides the name for the logging function. I'm not sure if we want to inject a function with some default name, so I made the whole thing optional.

File: lib/doctest.js [hunk 17]
```diff
@@ -197,20 +212,46 @@ function processLine(
...
       $test = accum.tests[accum.tests.length - 1];
-      $test[accum.state].value += '\n' + value;
-      (accum.state === INPUT ? appendToInput : appendToOutput) ($test);
-    } else if (accum.state === INPUT) {
+      $test[OUTPUT].values[$test[OUTPUT].values.length - 1].value += '\n'
+                                                                  + value;
+      appendToOutput ($test);
+    } else if ((match = MATCH_LOG.exec (trimmedLine)) != null) {
+      accum.state = LOG;
+      value = stripLeading (1, ' ', trimmedLine.replace (MATCH_LOG, ''));
```
Comment (davidchambers) [inline lib/doctest.js:235]: We could extract the remainder from `match`:

```diff
--- a/lib/doctest.js
+++ b/lib/doctest.js
@@ -181,4 +181,2 @@ var LOG = 'log';

-var MATCH_LOG = /^\[([a-zA-Z]+)\]:/;
-
 //  normalizeTest :: { output :: { value :: String } } -> Undefined
@@ -232,5 +230,5 @@ function processLine(
       appendToOutput ($test);
-    } else if ((match = MATCH_LOG.exec (trimmedLine)) != null) {
+    } else if ((match = /^\[([A-Za-z]+)\]:(.*)$/.exec (trimmedLine)) != null) {
       accum.state = LOG;
-      value = stripLeading (1, ' ', trimmedLine.replace (MATCH_LOG, ''));
+      value = stripLeading (1, ' ', match[2]);
       $test = accum.tests[accum.tests.length - 1];
```

Why do we restrict channel names to `[A-Za-z]+`?

File: test/index.js (hunk 13 above)
Comment (davidchambers) [reply test/index.js]: I pushed a commit which changes the formatting. :)

**Comment by davidchambers** [comment]
> Tests can use `log (channel) (value)` to produce additional output besides the return value of the expression:
>
> ```javascript
> > ( log ('stdout') (1)
> . , setTimeout (log ('stdout'), 1, 2)
> . , log ('stderr') (3) )
> [stdout]: 1
> [stderr]: 3
> undefined
> [stdout]: 2
> ```

Here is another interface worth considering:

```javascript
> ( stdout (1)
. , setTimeout (stdout, 1, 2)
. , stderr (3) )
[stdout]: 1
[stderr]: 3
undefined
[stdout]: 2
```

The basic usage would then be `log (f (x))` rather than `log ('log') (f (x))`. :)

The command-line interface would need to accept multiple log functions: `--log-function stdout --log-function stderr`.

File: lib/doctest.js (hunk 6 above)
Comment (Avaq) [reply lib/doctest.js:254]:
```js
[
  {channel: 'stdout', value: 1},
  {channel: 'stderr', value: 3},
  {channel: null, value: undefined},
  {channel: 'stdout', value: 2}
]
```

File: test/index.js (hunk 13 above)
Comment (Avaq) [reply test/index.js]: :+1:

**Comment by Avaq** [comment]
> The command-line interface would need to accept multiple log functions: `--log-function stdout --log-function stderr`.

Oh I like that idea a lot David!

File: lib/doctest.js (hunk 6 above)
Comment (davidchambers) [reply lib/doctest.js:254]: The line numbers are not necessarily correct. Consider this simple test case:

```console
$ printf "// %s\n" "> log ('log') ('hello'), 2 + 2" "[log]: 'hello'" "5" >/tmp/index.js

$ cat /tmp/index.js
// > log ('log') ('hello'), 2 + 2
// [log]: 'hello'
// 5

$ bin/doctest --log-function log -- /tmp/index.js
running doctests in /tmp/index.js...
.x
FAIL: expected 5 on line 2 (got 4)
```

The error message erroneously refers to line 2.

Comment (Avaq) [reply lib/doctest.js:254]: That's something I noticed as well. I think it's because the `loc` is still assigned to the outer structure, and the error message probably uses `loc.start`.

Comment (Avaq) [reply lib/doctest.js:254]: So for this feature to work properly, the structure would have to change from something like:

```js
{
  values: [{channel: 'a', value 1}],
  loc: {start: {line: 1, column: 1}, end: {line: 1, column: 2}}
}
```

To something like:

```js
[{
  channel: 'a',
  value 1,
  loc: {start: {line: 1, column: 1}, end: {line: 1, column: 2}}
}]
```

Comment (Avaq) [reply lib/doctest.js:254]: > The line numbers are not necessarily correct.

There might be a misunderstanding here. The values `1`, `3`, and `2` were not referring to line numbers, but to the values that were passed to `log`. The line numbers are tracked as a whole, as shown above, from the first line containing output, until the last line.

Comment (davidchambers) [reply lib/doctest.js:254]: > The line numbers are tracked as a whole, as shown above, from the first line containing output, until the last line.

Right. I don't find this behaviour intuitive, though. I would like to track the lines individually, as you proposed above. :)

**Comment by Avaq** [comment]
@davidchambers I addressed your two major points of feedback.
I find myself knowing this codebase like the back of my hand, at this point. :P

File: lib/doctest.js (hunk 7 above)
Comment (Avaq) [reply lib/doctest.js:369]: I think I'll call it out of scope for this PR. Got to draw the line somewhere.

**Comment by Avaq** [comment]
> Why do we restrict channel names to `[A-Za-z]+`?

The restriction at this point is that a channel name must be a [valid JavaScript identifier](https://mathiasbynens.be/notes/javascript-identifiers). Currently the `--log-line` parameter is overly permissive (any string), and the output assertion is overly restrictive (only letters). What do you think we should do?

**Comment by Avaq** [comment]
I just rebased, squashing some non-atomic commits. For the original history, see 0688bfbe8731288d2cffc872466a92fe6a88e8e3. I also pushed each new commit separately, to ensure that the tests pass on every commit. Let's see if there's any failing commits remaining.

**Comment by Avaq** [comment]
Despite my efforts to preserve backwards compatibility, there is at least one edge case where a user's code might break. That is, when they have code like this:

```js
// A good old regular test
//
// > 1
// 1

// [ACommentInBracketsForSomeReason]: Why?
// That's just how I like to format my comments.
```

Prior to my changes, this would work well. The reason is that the state machine would only parse lines starting with `// .` once it reached the OUTPUT state.

In the new situation, lines starting with `[chars]: ` are also valid once the state machine is in the OUTPUT state, because it needs to allow for asynchronous logs to occur after the expression output has been evaluated. This means that this code snippet that used to work will now generate a syntax error, as the code injector will inject `Why?` as the JavaScript to evaluate.

That is technically a breaking change.

----

> Could we simply replace `[A-Za-z]+` in the regular expression above with `.+`? -- https://github.com/davidchambers/doctest/pull/118#discussion_r438344109

Expanding the permitted characters in the brackets above will increase the odds for users to run into the issue I've described here. We could use a [strict JS identifier regex](https://stackoverflow.com/a/9392578/1001417), but that's a hassle.

I think the edge case would also be made much more rare if we implement your suggestion from here: https://github.com/davidchambers/doctest/pull/120#discussion_r338763923

File: lib/doctest.js [hunk 18]
```diff
@@ -241,14 +246,16 @@ function processLine(
...
     } else if (accum.state === INPUT || accum.state === LOG) {
-      accum.state = OUTPUT;
-      value = trimmedLine;
       $test = last (accum.tests);
-      $test[OUTPUT].push ({
-        channel: null,
-        value: value
-      });
-      output ($test);
+      if (!$test[OUTPUT].some (isStandardOutput)) {
```
Comment (Avaq) [inline lib/doctest.js:250]: This if statement actually describes another state in our state machine implicitly. It would not have to be there if we had two separate states for logging: `SYNC_LOG` and `ASYNC_LOG`. While the state machine is in `SYNC_LOG`, it allows transitions to the `OUTPUT` state, but while it's in `ASYNC_LOG` it only allows transitions to itself or `CLOSED`.

It might be nicer to implement it like that.

Comment (Avaq) [reply lib/doctest.js:250]: The valid state transitions would then be something like:

```
CLOSED -> CLOSED
CLOSED -> OPEN

OPEN -> INPUT

INPUT -> INPUT
INPUT -> OUTPUT
INPUT -> SYNC_LOG
INPUT -> ASYNC_LOG
INPUT -> CLOSED

OUTPUT -> OUTPUT
OUTPUT -> ASYNC_LOG
OUTPUT -> CLOSED

SYNC_LOG -> SYNC_LOG
SYNC_LOG -> OUTPUT
SYNC_LOG -> ASYNC_LOG
SYNC_LOG -> CLOSED

ASYNC_LOG -> ASYNC_LOG
ASYNC_LOG -> CLOSED
```

**Comment by Avaq** [comment]
> That is technically a breaking change.

I just took a shower and regained my sanity. The now blindingly obvious solution is to parse only the words provided on the CLI. It solves the "how should our regex be" issue, and regains our backwards compat: No `--log-function` arg means no log output parsing.
